controller/contour/deployment.go

```diff
@@ -0,0 +1,423 @@
+				DefaultMode: &cfgMapPerms,
+			},
+		},
+	}
+	cfgVolMt := corev1.VolumeMount{
+		Name:      cfgVolumeName,
+		MountPath: cfgVolumeMountDir,
+		ReadOnly:  true,
+	}
+	deploy.Spec.Template.Spec.Volumes = []corev1.Volume{certsVol, cfgVol}
+	deploy.Spec.Template.Spec.Containers[0].VolumeMounts = []corev1.VolumeMount{certsVolMt, cfgVolMt}
+
+	user := int64(65534)
+	group := int64(65534)
+	nonRoot := true
+	secCtx := corev1.PodSecurityContext{
+		RunAsUser:    &user,
+		RunAsGroup:   &group,
+		RunAsNonRoot: &nonRoot,
+	}
+	deploy.Spec.Template.Spec.SecurityContext = &secCtx
+
+	deploy.Spec.Template.Spec.Affinity = &corev1.Affinity{
+		PodAffinity: &corev1.PodAffinity{
+			PreferredDuringSchedulingIgnoredDuringExecution: []corev1.WeightedPodAffinityTerm{
+				{
+					Weight: int32(100),
+					PodAffinityTerm: corev1.PodAffinityTerm{
+						TopologyKey: "kubernetes.io/hostname",
+						LabelSelector: &metav1.LabelSelector{
+							MatchLabels: labels,
+						},
+					},
+				},
+			},
+		},
+	}
+	// TODO [danehans]: Readdress anti-affinity when https://github.com/projectcontour/contour/issues/2997
+	// is resolved.
+
+	return deploy, nil
+}
+
+// currentDeployment returns the current Deployment resource for the provided contour.
+func (r *Reconciler) currentDeployment(ctx context.Context, contour *operatorv1alpha1.Contour) (bool, *appsv1.Deployment, error) {
+	deploy := &appsv1.Deployment{}
+	key := types.NamespacedName{
+		Namespace: contour.Spec.Namespace.Name,
+		Name:      contour.Name,
+	}
+
+	if err := r.Client.Get(ctx, key, deploy); err != nil {
+		if errors.IsNotFound(err) {
+			return false, nil, nil
+		}
+		return false, nil, err
+	}
+
+	return true, deploy, nil
+}
+
+// createDeployment creates a Deployment resource for the provided deploy.
+func (r *Reconciler) createDeployment(ctx context.Context, deploy *appsv1.Deployment) error {
+	if err := r.Client.Create(ctx, deploy); err != nil {
+		return fmt.Errorf("failed to create deployment %s/%s: %v", deploy.Namespace, deploy.Name, err)
+	}
+	r.Log.Info("created deployment", "namespace", deploy.Namespace, "name", deploy.Name)
+
+	return nil
+}
+
+// updateDeployment updates a Deployment if current does not match desired.
+func (r *Reconciler) updateDeployment(ctx context.Context, current, desired *appsv1.Deployment) (bool, error) {
+	changed, updated := deploymentConfigChanged(current, desired)
+	if !changed {
+		return false, nil
+	}
+
+	if err := r.Client.Update(ctx, updated); err != nil {
+		return false, fmt.Errorf("failed to update deployment %s/%s: %v", updated.Namespace, updated.Name, err)
+	}
+	r.Log.Info("updated deployment", "namespace", updated.Namespace, "name", updated.Name)
+
+	return true, nil
+}
+
+// deploymentConfigChanged checks if current matches expected,
+// and if not returns the updated Deployment resource.
+func deploymentConfigChanged(current, expected *appsv1.Deployment) (bool, *appsv1.Deployment) {
```

Comment:
I'm not sure that I understand the need for all the checks here. Can we just compare the spec and the labels?

```Go
updated := current.DeepCopy()

if apiequality.Semantic.DeepEqual(current.Spec, expected.Spec) {
    changed = true
    expected.Spec.DeepCopyInto(&updated.Spec)
}

for _, l := range expected.Labels {
   // If label doesn't exist in `updated`, add it
}
```

Reply:
I originally went down this path because some spec fields were being defaulted by the api server instead of being specified by `desiredDeployment`. I'll update to compare spec and labels.